libs/ts-rest/fastify/src/lib/ts-rest-fastify.ts

```diff
@@ -0,0 +1,237 @@
+  [TKey in keyof T]: T[TKey] extends AppRouter
+    ? RecursiveRouterObj<T[TKey]>
+    : T[TKey] extends AppRoute
+    ? AppRouteImplementation<T[TKey]>
+    : never;
+};
+
+type RegisterRouterOptions = {
+  logInitialization?: boolean;
+  jsonQuery?: boolean;
+  responseValidation?: boolean;
+};
+
+const validateRequest = (
+  request: fastify.FastifyRequest,
+  reply: fastify.FastifyReply,
+  schema: AppRouteQuery | AppRouteMutation,
+  options: RegisterRouterOptions
+) => {
+  const paramsResult = checkZodSchema(request.params, schema.pathParams, {
+    passThroughExtraKeys: true,
+  });
+
+  const headersResult = checkZodSchema(request.headers, schema.headers, {
+    passThroughExtraKeys: true,
+  });
+
+  const queryResult = checkZodSchema(
+    options.jsonQuery
+      ? parseJsonQueryObject(request.query as Record<string, string>)
+      : request.query,
+    schema.query
+  );
+
+  const bodyResult = checkZodSchema(
+    request.body,
+    'body' in schema ? schema.body : null
+  );
+
+  if (
+    !paramsResult.success ||
+    !headersResult.success ||
+    !queryResult.success ||
+    !bodyResult.success
+  ) {
+    reply.status(400).send({
+      queryParameterErrors: queryResult.success ? null : queryResult.error,
+      pathParameterErrors: paramsResult.success ? null : paramsResult.error,
+      headerErrors: headersResult.success ? null : headersResult.error,
+      bodyErrors: bodyResult.success ? null : bodyResult.error,
+    });
+
+    return null;
+  }
+
+  return {
+    paramsResult,
+    headersResult,
+    queryResult,
+  };
+};
+
+export const initServer = () => ({
+  router: <T extends AppRouter>(router: T, args: RecursiveRouterObj<T>) => args,
+  registerRouter: <
+    T extends RecursiveRouterObj<TContract>,
+    TContract extends AppRouter
+  >(
+    contract: TContract,
+    routerImpl: T,
+    app: fastify.FastifyInstance,
+    options: RegisterRouterOptions = {
+      logInitialization: true,
+      jsonQuery: false,
+      responseValidation: false,
+    }
+  ) => {
+    recursivelyRegisterRouter(routerImpl, contract, [], app, options);
+  },
+});
+
+/**
+ *
+ * @param routeImpl - User's implementation of the route
+ * @param appRoute - the `ts-rest` contract for this route (e.g. with Path, params, query, body, etc.)
+ * @param fastify - the fastify instance to register the route on
+ */
+const registerRoute = <TAppRoute extends AppRoute>(
+  routeImpl: AppRouteImplementation<AppRoute>,
+  appRoute: TAppRoute,
+  fastify: fastify.FastifyInstance,
+  options: RegisterRouterOptions
+) => {
+  if (options.logInitialization) {
+    console.log(`[ts-rest] Initialized ${appRoute.method} ${appRoute.path}`);
+  }
+
+  fastify.route({
+    method: appRoute.method,
+    url: appRoute.path,
+    handler: async (request, reply) => {
+      const validationResults = validateRequest(
+        request,
+        reply,
+        appRoute,
+        options
+      );
+
+      if (validationResults === null) {
+        return;
+      }
```

Comment:
If we throw in `validateRequest` we won't need this check

Reply:
Good shout. I'll do the same in the express lib - unless you're doing that rewrite?

Reply:
I was thinking of doing it but it'd be a breaking change because you'd need to manually mount the middleware unless I can figure out a clean way to mount the new middleware by default and allow a user to override that. Let me see.

Reply:
@oliverbutler btw what I meant for this comment for to be combined with the approach in the comment above. I'm not sure what the behavior would be if you throw _anything_ but still return a response.

Anyway, you can check the approach I took here https://github.com/ts-rest/ts-rest/pull/232/commits/2d0ca77bbd8d2bee7e2a123bc5fdae6fc07ce89c

I followed the same format for the new `combined` error handler, but had to keep an old `default` one for backwards compatibility, but we won't need that problem for fastify. So in this case you can just either consume a custom error handler from the developer or use our own to return that 400 error with this new format.